File: Data Structures/Matching_Brackets/Matching_brackets.py
# python3

from collections import namedtuple

Bracket = namedtuple("Bracket", ["char", "position"])


def are_matching(left, right):
    return (left + right) in ["()", "[]", "{}"]


def find_mismatch(text):
    opening_brackets_stack = []
    for i, next in enumerate(text):
        if next in "([{": 
            opening_brackets_stack.append(Bracket(next,i))
        if next in ")]}":
            if not opening_brackets_stack or not are_matching(opening_brackets_stack.pop().char, next):
                return i+1

            
    if opening_brackets_stack:
        return opening_brackets_stack.pop().position +1
    else:
        return "Success"
            
                
         

def main():
    text = input()
    mismatch = find_mismatch(text)
    print(mismatch)
        


if __name__ == "__main__":
    main()

File: Data Structures/Matching_Brackets.cpp
#include <iostream>
#include <stack>
#include <string>

struct Bracket {
    char type;
    int position;
    Bracket(char type, int position):
        type(type),
        position(position)
    {}

    bool Matchc(char c) {
        if (type == '[' && c == ']')
            return true;
        if (type == '{' && c == '}')
            return true;
        if (type == '(' && c == ')')
            return true;
        return false;
    }
};

int main() {
    std::string text;
    getline(std::cin, text);

    std::stack <Bracket> opening_brackets_stack;
    for (int unsigned position = 0; position < text.length(); ++position) {
        char next = text[position];

        if (next == '(' || next == '[' || next == '{') {
            opening_brackets_stack.push(Bracket(next,position+1));
        }

        if (next == ')' || next == ']' || next == '}') {
            if(!opening_brackets_stack.empty() && opening_brackets_stack.top().Matchc(next)){
                opening_brackets_stack.pop();
            }
            else{
                std::cout<<position+1<<std::endl;
                return 0;
            };
        }        
    }
    if(!opening_brackets_stack.empty()){
        std::cout<<opening_brackets_stack.top().position<<std::endl;
    }
    else{
        std::cout<<"Success"<<std::endl;
    }
    return 0;
}